stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import GCN, GAT, APPNP, PPNP
+from ..core.experimental import experimental
+
+from tensorflow.keras.layers import Input, Lambda, Layer, GlobalAveragePooling1D
+import tensorflow as tf
+from tensorflow.keras import backend as K
+import warnings
+
+__all__ = ["DeepGraphInfoMax", "BilinearTransform"]
+
+
+class BilinearTransform(Layer):
+    """
+    This Layer computes the Discriminator function for Deep Graph Infomax (https://arxiv.org/pdf/1809.10341.pdf).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shapes):
+
+        self.kernel = self.add_weight(
+            shape=(input_shapes[0][-1], input_shapes[0][-1]),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list or tuple of tensors with shapes [(N, F), (F,)] containing the node features and summary feature
+                vector.
+        """
+
+        features, summary = inputs
+
+        summary_rank = len(summary.shape)
+        feat_rank = len(features.shape)
+
+        score = tf.tensordot(self.kernel, summary, [[1], [summary_rank - 1]])
```

Comment:
(The docs for `tensordot` are ... really confusing, and seem to have a pile of typos. 😦 )

I take it you switched to this more general operator because the tensors now have the batch dimension, and so `tf.matvec` no longer works?

Does it work if `self.kernel` has a batch dimension added?

```suggestion
        score = tf.matvec(tf.expand_dims(self.kernel, axis=0), summary)
```

or even just add a `1` to the `add_weight`: `shape=(1, ...)`.

Reply:
(More like the docs for `tensorflow` ... 😕 )

Actually I never tested out just using `matvec`, just tried it and it works fine! I had some other issues with keeping the batch dimension and prematurely blamed `matvec` and swapped it out for `tensordot`. I've put `matvec` back in and things are simple again :)